cloudify_gcp/utils.py

```diff
@@ -0,0 +1,466 @@
+                        props.pop(key, None)
+                    if get:
+                        props.update(self.get())
+                else:
+                    raise NonRecoverableError(
+                            'Unknown status response from operation')
+
+            else:
+                # Actually run the method
+                response = func(self, *args, **kwargs)
+                ctx.instance.runtime_properties['_operation'] = response
+                ctx.operation.retry('Operation started')
+
+        return wraps(func)(wrapper)
+    return decorator
+
+
+def retry_on_failure(msg, delay=constants.RETRY_DEFAULT_DELAY):
+    def _retry_on_failure(func):
+        def _decorator(*args, **kwargs):
+            try:
+                return func(*args, **kwargs)
+            except HttpError as error:
+                if is_resource_used_error(error):
+                    ctx.operation.retry(msg, delay)
+                else:
+                    raise error
+
+        return wraps(func)(_decorator)
+
+    return _retry_on_failure
+
+
+def throw_cloudify_exceptions(func):
+    def _decorator(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except GCPError as e:
+            raise NonRecoverableError(e.message)
+
+    return wraps(func)(_decorator)
+
+
+def get_gcp_config():
+    def _get_gcp_config_from_properties():
+        try:
+            return ctx.node.properties[constants.GCP_CONFIG]
+        except NonRecoverableError:
+            return ctx.source.node.properties[constants.GCP_CONFIG]
+
+    gcp_config_from_properties = _get_gcp_config_from_properties()
+    if gcp_config_from_properties:
+        gcp_config = gcp_config_from_properties
+    else:
+        try:
+            config = ctx.provider_context['resources'][constants.GCP_CONFIG]
+        except KeyError:
+            try:
+                with open(expanduser(constants.GCP_DEFAULT_CONFIG_PATH)) as f:
+                    gcp_config = yaml.load(f)
+            except OSError:
+                raise NonRecoverableError(
+                    '{} not provided as a property. '
+                    'and the provider context '
+                    'is not set up either'.format(constants.GCP_CONFIG))
+        gcp_config = deepcopy(config)
+
+    return update_zone(gcp_config)
+
+
+def update_zone(gcp_config):
+    def _get_zone_from_runtime_properties():
+        try:
+            return ctx.instance.runtime_properties.get(constants.GCP_ZONE)
+        except NonRecoverableError:
+            src = ctx.source.instance.runtime_properties
+            tar = ctx.target.instance.runtime_properties
+            return src.get(constants.GCP_ZONE) or tar.get(constants.GCP_ZONE)
+
+    non_default_zone = _get_zone_from_runtime_properties()
+    if non_default_zone:
+        gcp_config['zone'] = non_default_zone
+
+    return gcp_config
+
+
+def get_manager_provider_config():
+    provider_config = ctx.provider_context.get('resources', {})
```

Comment:
I think this is still needed because `compute.instance.add_to_security_groups` uses the provider context to get the agents security group